File: AudioKit/Common/Internals/Helpers/ExceptionCatcher.m
//
//  ExceptionCatcher.m
//  AudioKit
//
//  Created by Stéphane Peter, revision history on Github.
//  Copyright © 2018 AudioKit. All rights reserved.
//

#import "ExceptionCatcher.h"

void AKTryOperation(void (^ _Nonnull tryBlock)(void),
                    void (^ _Nullable catchBlock)(NSException * _Nonnull))
{
    @try {
        tryBlock();
    }
    @catch (NSException *exception) {
        if (catchBlock)
            catchBlock(exception);
    }
}

File: AudioKit/Common/Internals/Helpers/ExceptionCatcher.swift
//
//  ExceptionCatcher.swift
//  AudioKit
//
//  Created by Stéphane Peter, revision history on Github.
//  Copyright © 2018 AudioKit. All rights reserved.
//

import Foundation

public func AKTry(_ operation: @escaping (() throws -> Void)) throws
{
    var error: Error?
    
    let theTry = {
        do {
            try operation()
        } catch let ex {
            error = ex
        }
    }
    
    let theCatch: (NSException) -> Void = { except in
        error = NSError(domain: "io.audiokit.AudioKit",
                        code: 0,
                        userInfo: ["Exception": except.userInfo ?? except.name])
    }
    
    AKTryOperation(theTry, theCatch)
    
    if let error = error { // Caught an exception
        throw error
    }
}
